File: archived/core/ed25519/python/nonce_cache.py
"""
PEAC Protocol v0.9.1
Nonce cache for anti-replay (memory only)
Apache 2.0 License
"""

import time

NONCE_TTL = 5 * 60  # 5 minutes in seconds

class NonceCache:
    def __init__(self):
        self.nonces = {}

    def add(self, nonce: str, timestamp: int) -> bool:
        now = int(time.time() * 1000)
        # timestamp in ms
        if nonce in self.nonces:
            return False
        if abs(now - int(timestamp)) > NONCE_TTL * 1000:
            return False
        self.nonces[nonce] = now
        # Schedule cleanup not needed in scripts, but in prod: clean old entries periodically
        return True

    def clear(self):
        self.nonces = {}

File: archived/core/ed25519/node/nonceCache.js
/**
 * PEAC Protocol v0.9.1
 * Nonce cache for anti-replay (memory-only, not persistent)
 * Apache 2.0 License
 */

const NONCE_TTL = 5 * 60 * 1000; // 5 minutes in ms

class NonceCache {
  constructor() {
    this.nonces = new Map();
  }

  // Returns true if nonce is new and valid; false if replayed or expired
  add(nonce, timestamp) {
    const now = Date.now();
    if (this.nonces.has(nonce)) return false;
    if (Math.abs(now - Number(timestamp)) > NONCE_TTL) return false;
    this.nonces.set(nonce, now);
    // Schedule cleanup
    setTimeout(() => this.nonces.delete(nonce), NONCE_TTL);
    return true;
  }

  // For tests/debug
  clear() {
    this.nonces.clear();
  }
}

module.exports = NonceCache;
